### PR TITLE
Fix EarlyStopping logic when min_epochs not met

### LINKED ISSUE #6699: Early stopping skips batches when min_epochs not reached
## 🐛 Bug
When EarlyStopping hits but `min_epochs` is not reached, it seems that only the first batch of an epoch is used.

TL;DR: In a setting of 100 samples, batchsize 25 (=4 batches), `training_step()` should always be called exactly 4 times. When EarlyStopping hit before `min_epochs` is reached, it is only called once.

### To Reproduce

I used the [**original BoringModel template**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and changed: 

- `num_samples = 100`
- `batch_size=25` in train, val and test
- Add ` print('train, batch=', batch_idx)` to `self.training_step()`
- Add EarlyStopping and min_epochs when initializing the trainer

Resulting in: 
https://colab.research.google.com/drive/11tlIU9NusGPeXJLUKA52ECuhIXOCH_4k?usp=sharing

I added the relevant output when training as image here: 
image

### Expected behavior
In my BoringModel, `self.training_step()` must be called 4 times in each epoch as long as min_epochs is not reached. Otherwise I suspect that the remaining 3 batches were not used to update the model parameters. 

### Environment
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.0+cu101
	- pytorch-lightning: 1.2.5
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### Additional context
- I took a look into the sourcecode during my actual analysis and found that in `pytorch_lightning/trainer/training_loop.py`:`TrainLoop.run_training_epoch()`, `self.trainer.should_stop` is True (and breaks after the first batch) when EarlyStopping hit. 

- As a result, it seems that once EarlyStopping hit before min_epochs is reached, it always stops with min_epochs, even if the stopping-criteria is not meet then. (But I did not MWE that.)

- PS: Is it intended that `BoringModel.forward()` is *not* called in the template but `self.layer()` instead?

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6699

When EarlyStopping is triggered but trainer.min_epochs is not satisfied, training must be forced to continue.
This bugfix prevents breaking out of the epoch loop too early in this case.

**Test fails on master.**

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -546,6 +547,41 @@ def test_trainer_min_steps_and_epochs(tmpdir):
...
+
+    class TestModel(BoringModel):
+        training_step_invoked = 0
+
+        def training_step(self, batch, batch_idx):
+            output = super().training_step(batch, batch_idx)
+            output["loss"] = output["loss"] * 0.0  # force minimal loss to trigger early stopping
+            self.log("loss", output["loss"])
+            self.training_step_invoked += 1
+            # print(batch_idx, self.trainer.current_epoch)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:561]:
```suggestion
```

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/test_trainer.py:561]: thanks!

**Review by ananthsub — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -546,6 +547,40 @@ def test_trainer_min_steps_and_epochs(tmpdir):
     assert trainer.global_step >= math.floor(num_train_samples * 1.5), "Model did not train for at least min_steps"
 
 
+def test_trainer_min_steps_and_min_epochs_not_reached(tmpdir, caplog):
```
Comment (Borda) [inline tests/trainer/test_trainer.py:550]: shall we rather integrate it in an exiting test?
CC: @carmocca

**Review by Borda — COMMENTED** [review]
LGTM

**Review by Borda — APPROVED** [review]

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (awaelchli) [reply tests/trainer/test_trainer.py:550]: hmm, it doesn't fit well with the other test. I don't see an easy way because we need a particular set of trainer arguments.

**Review by akihironitta — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
